source/image/ImageBuffer.cpp

```diff
@@ -359,6 +379,136 @@ namespace {
+				return 0;
+			}
+			if(frameTimeUnit == -1 || (frameTimeUnit > timing.duration && timing.duration))
+				frameTimeUnit = timing.duration;
+		}
+		// Based on this unit, we can calculate how many times each frame is repeated.
+		vector<size_t> repeats(decoder->imageCount);
+		size_t bufferFrameCount = 0;
+		for(size_t i = 0; i < static_cast<size_t>(decoder->imageCount); ++i)
+		{
+			result = avifDecoderNthImageTiming(decoder.get(), 0, &timing);
+			if(result != AVIF_RESULT_OK)
+			{
+				Logger::LogError("Could not get image timing for \"" + path.generic_string() + "\": " + avifResultToString(result));
+				return 0;
+			}
+			repeats[i] = round(timing.duration / frameTimeUnit);
+			bufferFrameCount += repeats[i];
+		}
+
+		// Now that we know the buffer's frame count, we can allocate the memory for it.
+		// If this is an image sequence, the preconfigured frame count is wrong.
+		try {
+			if(bufferFrameCount > 1)
+				buffer.Clear(bufferFrameCount);
+			buffer.Allocate(decoder->image->width, decoder->image->height);
+		}
+		catch(const bad_alloc &)
+		{
+			const string message = "Failed to allocate contiguous memory for \"" + path.generic_string() + "\"";
+			Logger::LogError(message);
+			throw runtime_error(message);
+		}
+		if(static_cast<unsigned>(buffer.Width()) != decoder->image->width
+				|| static_cast<unsigned>(buffer.Height()) != decoder->image->height)
+		{
+			Logger::LogError("Invalid dimensions for \"" + path.generic_string() + "\"");
+			return 0;
+		}
+
+		// Load each image in the sequence.
+		int avifFrameIndex = 0;
+		size_t bufferFrame = 0;
+		while(avifDecoderNextImage(decoder.get()) == AVIF_RESULT_OK)
+		{
+			// Ignore frames with insufficient duration.
+			if(!repeats[avifFrameIndex])
+				continue;
+
+			avifRGBImage image;
+			avifRGBImageSetDefaults(&image, decoder->image);
+			image.depth = 8; // force 8-bit color depth
+			image.alphaPremultiplied = alphaPreMultiplied;
+			image.rowBytes = image.width * avifRGBImagePixelSize(&image);
+			image.pixels = reinterpret_cast<uint8_t *>(buffer.Begin(0, frame + bufferFrame));
+
+			result = avifImageYUVToRGB(decoder->image, &image);
+			if(result != AVIF_RESULT_OK)
+			{
+				Logger::LogError("Conversion from YUV failed for \"" + path.generic_string() + "\": " + avifResultToString(result));
+				return bufferFrame;
+			}
+
+			// Now copy the image in the buffer to match frame timings.
+			for(size_t i = 1; i < repeats[avifFrameIndex]; ++i)
+			{
+				uint8_t *end = reinterpret_cast<uint8_t *>(buffer.Begin(0, frame + bufferFrame + 1));
+				uint8_t *dest = reinterpret_cast<uint8_t *>(buffer.Begin(0, frame + bufferFrame + i));
+				std::copy(image.pixels, end, dest);
+			}
+			bufferFrame += repeats[avifFrameIndex];
+
+			avifFrameIndex++;
+			avifRGBImageSetDefaults(&image, decoder->image);
+		}
+
+		if(avifFrameIndex != decoder->imageCount || bufferFrame != bufferFrameCount)
```

Comment:
`bufferFrameCount` sums `repeats` values, and in this loop we skip only the entries that have `repeats` == 0, so I don't think we need this, but do as you prefer.
```suggestion
		if(avifFrameIndex != decoder->imageCount)
```

Reply:
I know, but extra checks can't hurt.